File: AuLiteSql/src/main/java/com/au/aulitesql/BaseEntityListDao.java
package com.au.aulitesql;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import com.au.aulitesql.actions.ICallback;

import java.util.List;

public abstract class BaseEntityListDao<E extends Entity> {
    public final Class<E> entityClass;
    public BaseEntityListDao(Class<E> clazz) {
        this.entityClass = clazz;
    }

    ////////////load////////////////
    @WorkerThread
    public void loadAllFilter(String fieldName, Object value) {
        loadAllFilter(fieldName, value, null);
    }
    @WorkerThread
    public abstract void loadAllFilter(String fieldName, Object value, ICallback<List<E>> callback);

    @WorkerThread
    public void loadAll() {
        loadAll(null);
    }
    @WorkerThread
    public abstract void loadAll(ICallback<List<E>> callback);

    @WorkerThread
    public void loadAll(@NonNull String fieldName, Object value,
                                 String groupBy, String having, String orderBy) {
        loadAll(fieldName, value, groupBy, having, orderBy, null);
    }

    @WorkerThread
    public abstract void loadAll(@NonNull String fieldName, Object value,
                                       String groupBy, String having, String orderBy, ICallback<List<E>> callback);

    @WorkerThread
    public void loadAll(@NonNull String selections, @NonNull String[] selectionArgs,
                                 String groupBy, String having, String orderBy) {
        loadAll(selections, selectionArgs, groupBy, having, orderBy, null);
    }

    @WorkerThread
    public abstract void loadAll(@NonNull String selections, @NonNull String[] selectionArgs,
                                       String groupBy, String having, String orderBy, ICallback<List<E>> callback);

    @WorkerThread
    public void rawLoadAll(String sql, String[] selectionArgs) {
        rawLoadAll(sql, selectionArgs, null);
    }

    @WorkerThread
    public abstract void rawLoadAll(String sql, String[] selectionArgs, ICallback<List<E>> callback);

    //////////////////delete////////////////

    @WorkerThread
    public void deleteAll(List<E> dataList) {
        deleteAll(dataList, null);
    }

    @WorkerThread
    public abstract void deleteAll(List<E> dataList, ICallback<Integer> deleteCountCallback);

    @WorkerThread
    public void delete(E instance) {
        delete(instance, null);
    }

    @WorkerThread
    public abstract void delete(E instance, ICallback<Boolean> deleteSuccessCallback);

    @WorkerThread
    public abstract void clear(ICallback<Boolean> clearSuccessCallback);

    ///////////////////save/////////////////
    @WorkerThread
    public void save(E instance) {
        save(instance, null);
    }

    @WorkerThread
    public abstract void save(E instance, ICallback<Boolean> saveSuccessCallback);

    @WorkerThread
    public void saveAll(List<E> dataList) {
        saveAll(dataList, null);
    }

    @WorkerThread
    public abstract void saveAll(List<E> dataList, ICallback<Integer> saveSuccessCountCallback);

    //////////////////////////////////
    @WorkerThread
    public void update(E instance) {
        update(instance, null);
    }

    @WorkerThread
    public void update(E instance, @Nullable ICallback<Boolean> updateSuccessCallback) {
        var id = saveInstanceToDb(instance);
        if(updateSuccessCallback != null) updateSuccessCallback.callback(id >= 0L);
    }

    /**
     * 增加或者修改。
     */
    @WorkerThread
    public static long saveInstanceToDb(Entity entity) {
        AuLiteSql.getDao().save(entity);
        return entity.getId();
    }

    //////////////////////////////////
    //////////////////////////////////
    //////////////////////////////////
}

File: AuLiteSqlKt/src/main/java/com/au/aulitesqlkt/BaseEntityListDao.kt
package com.au.aulitesqlkt

import com.au.aulitesql.Entity
import com.au.aulitesql.actions.ICallback

abstract class BaseEntityListDao<E : Entity?>(val entityClass: Class<E>) {
    ////////////load////////////////
    fun loadAllFilter(fieldName: String?, value: Any?) {
        loadAllFilter(fieldName, value, null)
    }

    abstract fun loadAllFilter(fieldName: String?, value: Any?, callback: ICallback<List<E>?>?)
    fun loadAll() {
        loadAll(null)
    }

    abstract fun loadAll(callback: ICallback<List<E>?>?)
    fun loadAll(
        fieldName: String, value: Any?,
        groupBy: String?, having: String?, orderBy: String?
    ) {
        loadAll(fieldName, value, groupBy, having, orderBy, null)
    }

    abstract fun loadAll(
        fieldName: String, value: Any?,
        groupBy: String?, having: String?, orderBy: String?, callback: ICallback<List<E>?>?
    )

    fun loadAll(
        selections: String, selectionArgs: Array<String?>,
        groupBy: String?, having: String?, orderBy: String?
    ) {
        loadAll(selections, selectionArgs, groupBy, having, orderBy, null)
    }

    abstract fun loadAll(
        selections: String, selectionArgs: Array<String?>,
        groupBy: String?, having: String?, orderBy: String?, callback: ICallback<List<E>?>?
    )

    fun rawLoadAll(sql: String?, selectionArgs: Array<String?>?) {
        rawLoadAll(sql, selectionArgs, null)
    }

    abstract fun rawLoadAll(sql: String?, selectionArgs: Array<String?>?, callback: ICallback<List<E>?>?)

    //////////////////delete////////////////
    fun deleteAll(dataList: List<E>?) {
        deleteAll(dataList, null)
    }

    abstract fun deleteAll(dataList: List<E>?, deleteCountCallback: ICallback<Int?>?)
    fun delete(instance: E) {
        delete(instance, null)
    }

    abstract fun delete(instance: E, deleteSuccessCallback: ICallback<Boolean?>?)
    abstract fun clear(clearSuccessCallback: ICallback<Boolean?>?)

    ///////////////////save/////////////////
    fun save(instance: E) {
        save(instance, null)
    }

    abstract fun save(instance: E, saveSuccessCallback: ICallback<Boolean?>?)
    fun saveAll(dataList: List<E>?) {
        saveAll(dataList, null)
    }

    abstract fun saveAll(dataList: List<E>?, saveSuccessCountCallback: ICallback<Int?>?) //////////////////////////////////
    //////////////////////////////////
    //////////////////////////////////
    //////////////////////////////////
}